### DISCUSSION TITLE
Map different collections based on association object field

### DISCUSSION THREAD

**Discussion by lesinigo** [sqlalchemy/sqlalchemy#11445] [Usage Questions]
I have a use case where some `Child` elements are associated in different ways to `Parent` elements and could be represented with [the Association Object pattern](https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#association-object) by adding the relation "kind" to the association object.

I'd like to leverage SQLAlchemy mappings to keep the code that uses this model neat and simple, possibly having different mapped collections on my `Parent` class that are both readable and writable, just like the `Parent.children` mapping on the Association Object example but I'd like to differentiate various children groups based on the value of a column in the association table.

My current best implementation follows this simplified but working example (SQLAlchemy 2.0.30, Python 3.11):

```python
class Child(Base):
    __tablename__ = "children"
    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, insert_default=uuid.uuid4)
    data: Mapped[str] = mapped_column()

class GreenAssociation(Base):
    __tablename__ = "associations"
    parent_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("parents.id"), primary_key=True)
    parent: "Mapped[Parent]" = relationship(back_populates="green_children_associations")
    type: Mapped[str] = mapped_column(primary_key=True)
    child_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("children.id"), primary_key=True)
    child: Mapped[Child] = relationship("Child", cascade="all, delete")
    __mapper_args__ = {"polymorphic_on": "type", "polymorphic_identity": "green"}

class BlueAssociation(GreenAssociation):
    parent: "Mapped[Parent]" = relationship(back_populates="blue_children_associations")
    __mapper_args__ = {"polymorphic_identity": "blue"}

class Parent(Base):
    __tablename__ = "parents"
    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, insert_default=uuid.uuid4)
    green_children_associations: Mapped[list[GreenAssociation]] = relationship(
        primaryjoin="and_(Parent.id==GreenAssociation.parent_id, GreenAssociation.type=='green')",
        back_populates="parent",
        cascade="all, delete-orphan",
        overlaps="parent",
    )
    blue_children_associations: Mapped[list[BlueAssociation]] = relationship(
        primaryjoin="and_(Parent.id==BlueAssociation.parent_id, BlueAssociation.type=='blue')",
        back_populates="parent",
        cascade="all, delete-orphan",
        overlaps="parent",
    )
    green_children: "AssociationProxy[list[Child]]" = association_proxy(
        "green_children_associations",
        "child",
        creator=lambda v: GreenAssociation(type="green", child=v),
    )
    blue_children: "AssociationProxy[list[Child]]" = association_proxy(
        "blue_children_associations",
        "child",
        creator=lambda v: BlueAssociation(type="blue", child=v),
    )
```

As far as I know the `green_children` and `blue_children` proxies are working correctly to read and modify the children collections. I had to "force" the association types in both `relationship(primaryjoin="...")` and in `association_proxy(creator=...)` even though, in theory, SQLAlchemy should have already done it due to the inherited classes.

A problem that comes up with this kind of mapping is that I still get a warning about overlapping relationships:

> SAWarning: relationship 'Parent.blue_children_associations' will copy column parents.id to column associations.parent_id, which conflicts with relationship(s): 'Parent.green_children_associations' (copies parents.id to associations.parent_id). If this is not the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially overlapping, the orm.foreign() annotation can be used to isolate the columns that should be written towards.   To silence this warning, add the parameter 'overlaps="green_children_associations"' to the 'Parent.blue_children_associations' relationship. (Background on this warning at: https://sqlalche.me/e/20/qzyx) (This warning originated from the `configure_mappers()` process, which was invoked automatically in response to a user-initiated operation.)

I could get rid of the warning by specifying `overlaps=...` on every relationship, but with non-trivial number of "association types" (think ~10) it becomes ugly real quick to write all possible overlaps.

Am I missing something here or there is some better way to achieve the same results?

**Comment by zzzeek** [comment]
the warning there is strictly to let you know that your code is capable of putting data inside of green_children_associations and blue_children_associations that can in fact be in conflict.   you could put a BlueAssociation into the green_children_associations collection of one Parent, and into the blue_children_associations collection of another, and only one of them would "win" at flush time.

assuming your code makes sure this doesnt happen, that's what the overlaps is for.    Make a function that generates the `relationship()` you want with the options you need intact.

**Comment by lesinigo** [comment]
If I am understanding correctly the warning is basically telling me that I could do something like this and mess up everything:

```python
    green_children: "AssociationProxy[list[Child]]" = association_proxy(
        "green_children_associations",
        "child",
        creator=lambda v: GreenAssociation(type="blue", child=v),   # <--- "blue" is wrong here
    )
```

Since I'm actually using [multiple classes with single table inheritance](https://docs.sqlalchemy.org/en/20/orm/inheritance.html#relationships-with-single-table-inheritance), SQLAlchemy could theoretically know better and allow me to configure mappings where I work on the collections of final entities (the `Child`s in my example above) without dealing with associations objects at all and that should avoid any possibility of conflicting data.

Am I correct in assuming that this is some weird situation not actually possible in current SA and that a solution like my example above is the best possible one? Would anyone solve the use case with a different/better implementation?

**Reply by zzzeek** [reply]
I dont know you can do `GreenAssociation(type='blue')`, just that you can put a `BlueAssociation` into `Parent.green_children_associations` directly, since a `BlueAssociation` is a `GreenAssociation` the way you have it there.     if your code only uses the association proxy for mutation, then yes that would prevent that from happening.

I think what could improve your mapping would be if GreenAssociation and BlueAssociaiton were sibling classes, rather than parent/child.    A common base with `__abstract__` can have what they share.    I dont think you even need the special primaryjoin conditions for this either, SQLAlchemy should add that criteria automatically (though I could be mistaken).
